### PR TITLE
feat: Creates task to calculate metrics from K562 perturbation data

### LINKED ISSUE #342: Perturb Seq Task: Initial Implementation
## Inputs

- A matrix with dimensions of cells x genes (NV not model embedding), and information (a list?) of positions that were masked. To accelerate development purposes, will start with an h5ad file containing random model predictions provided by NVIDIA team. 
- Coordinates of masked positions – TBD how this is stored
- Differential expression related values on input dataset and background matched controls. It is believed these files can be included with the dataset (to be validated during task development). To accelerate development, they will initially be side loaded.

## Objective
Evaluate model's ability to predict expression of masked, perturbed genes in a cell when a specified percentages of other perturbed genes are masked. 

## Outputs
Predicted expression values, and corresponding actual expression values

## Risk Factors
May require changes to cache format and/or dataset class to handle inputs and coordinates of masked data

## Sub-tasks

- Create initial implementation of task. Test will initially use DE values determined by Wilcox test.
- Does DE prediction need to be run on model predictions to determine some of the  metrics? [I think the answer to this is NO and that the accuracy metrics are from a binary version of up/down prediction.]
- Update basic (non end to end) tests for task

### LINKED ISSUE #343: Perturb Seq Task: Update Cache Format
Update caching functionality if needed, e.g. to allow for masked coordinates. Update task for compatibility with cached npz matrices. Requires converting sample dataset to npz matrix.

### LINKED ISSUE #344: Perturb Seq Task: Add Additional DE Methods
Add compatibility with additional DE tests (T-test, union, intersection) for models which output data with negative or zero values, e.g. z-score normalized. Update tests as needed

### LINKED ISSUE #345: Perturb Seq Task: Update Cached Inputs in Task
Update task to remove side-loading of DE and matched control data. Requires prior updates to dataset.

### LINKED ISSUE #346: Perturb Seq Metric: Initial Metric Implementation
Add metric class which calculates expected metric results – spearman correlation of values and binary prediction accuracy metrics. Most, if not all, of these metrics exist in registry and are used for other metric classes. Spearman R might need to be implemented. Update metric tests if needed.

### LINKED ISSUE #366: Perturb Seq Task: Implement Baseline
Design and implement a baseline for the Perturb seq task.

### PR DESCRIPTION
**Pull request by polinabinder1**
So far, this uses data from Wilcoxon tests to identify differentially expressed genes. Additionally, it relies on user-set file names.

closes #342 
closes #343 
closes #344 
closes #345
closes #346 
closes #366

### CODE REVIEW

File: src/czbenchmarks/tasks/single_cell/k562_perturbation.py [hunk 1]
```diff
@@ -0,0 +1,288 @@
...
+from ...metrics import metrics_registry
+from ...metrics.types import MetricResult, MetricType
+from ...constants import RANDOM_SEED
+
+from tqdm import tqdm
+
+logger = logging.getLogger(__name__)
+
+
+def binarize_values(y_true, y_pred):
```
Comment (mlgill) [inline src/czbenchmarks/tasks/single_cell/k562_perturbation.py:16]: This could go into a shared utils folder if it is likely to be used in other tasks.

File: src/czbenchmarks/tasks/single_cell/k562_perturbation.py [hunk 2]
```diff
@@ -0,0 +1,288 @@
...
+    pred_binary = (y_pred > 0).astype(int)
+    true_binary = (y_true > 0).astype(int)
+    return true_binary, pred_binary
+
+
+class K562PerturbationTaskInput(TaskInput):
+    """Pydantic model for PerturbationTask inputs."""
+
+    de_res_wilcoxon_df: pd.DataFrame
+    pred_df: pd.DataFrame
```
Comment (mlgill) [inline src/czbenchmarks/tasks/single_cell/k562_perturbation.py:29]: I think this should use the CellRepresentation type, but we may want to be more specific as I don't understand how the coordinates of the predicted values are being determined here. Let's discuss in standup today.

File: src/czbenchmarks/tasks/single_cell/k562_perturbation.py [hunk 3]
```diff
@@ -0,0 +1,288 @@
...
+
+
+class K562PerturbationOutput(TaskOutput):
+    """Output for perturbation task."""
+
+    pred_log_fc_dict: Dict[str, np.ndarray]
+    true_log_fc_dict: Dict[str, np.ndarray]
+
+
+class K562PerturbationTask(Task):
```
Comment (mlgill) [inline src/czbenchmarks/tasks/single_cell/k562_perturbation.py:39]: This task will be used on other datasets, so the name shouldn't be specific to Replogle or K562 cells. Maybe something like "PerturbationExpressionPrediction"? Or something even better than you think of. Will need to update the pydantic classes, example script, etc.

File: src/czbenchmarks/tasks/single_cell/k562_perturbation.py [hunk 4]
```diff
@@ -0,0 +1,288 @@
...
+                task_input.de_res_wilcoxon_df["target_gene"] == condition
+            ]
+            if len(condition_de_df) < self.min_de_genes:
+                continue
+
+            cell_condition = cell_representation[
+                cell_representation.index.str.endswith(f"_{condition}")
+            ]
+            condition_cells = cell_condition[
+                cell_condition["gene"] != "non-targeting"
```
Comment (mlgill) [inline src/czbenchmarks/tasks/single_cell/k562_perturbation.py:94]: It will be a minor update in the future, but will be getting the name for the control condition from the dataset, since it can change by dataset.

**Review by mlgill — COMMENTED** [review]
This is a great start, have left some feedback that we can discuss today.

File: src/czbenchmarks/tasks/single_cell/k562_perturbation.py [hunk 5]
```diff
@@ -0,0 +1,288 @@
...
+    y_pred = y_pred[ids]
+    pred_binary = (y_pred > 0).astype(int)
+    true_binary = (y_true > 0).astype(int)
+    return true_binary, pred_binary
+
+
+class K562PerturbationTaskInput(TaskInput):
+    """Pydantic model for PerturbationTask inputs."""
+
+    de_res_wilcoxon_df: pd.DataFrame
```
Comment (mlgill) [inline src/czbenchmarks/tasks/single_cell/k562_perturbation.py:28]: We will probably need to take both wilcoxon and t-test, so suggest simplifying the name. And there will need to be a parameter as to which was used.

File: src/czbenchmarks/metrics/implementations.py [hunk 6]
```diff
@@ -146,3 +150,43 @@
     default_params={"metric": "auroc", "classifier": None},
     tags={"label_prediction"},
 )
+
+metrics_registry.register(
```
[the remark this answers is not in the document]
Comment (polinabinder1) [reply src/czbenchmarks/metrics/implementations.py:154]: added test cases

File: src/czbenchmarks/tasks/single_cell/k562_perturbation.py (hunk 4 above)
Comment (polinabinder1) [reply src/czbenchmarks/tasks/single_cell/k562_perturbation.py:94]: fixed as a parameter.

File: examples/test_equivalency_perturbation_task.py [hunk 7]
```diff
@@ -0,0 +1,388 @@
+# This tests that the perturbation prediection task with the new data formats matches
+# what is outputted in Maria's intial notebook.
```
Comment (mlgill) [inline examples/test_equivalency_perturbation_task.py:2]: @polinabinder1 Is it correct that this is for internal validation only and will not be part of actual PR?

File: tests/tasks/test_tasks.py [hunk 8]
```diff
@@ -325,3 +328,286 @@ def test_perturbation_task():
...
+        assert len(results) == num_metric_types
+
+        # Each list should contain MetricResult objects
+        for metric_list in results:
+            assert isinstance(metric_list, list)
+            assert all(isinstance(r, MetricResult) for r in metric_list)
+            # Should have results for at least one condition (gene_A or gene_B)
+            # Make this more flexible - if no conditions pass filtering, that's okay for a test
+            if len(metric_list) == 0:
+                print("Warning: No results generated for this metric type")
```
Comment (mlgill) [inline tests/tasks/test_tasks.py:581]: Could use logging if you want, although I think this is optional since it's a test.

File: src/czbenchmarks/tasks/utils.py [hunk 9]
```diff
@@ -25,7 +25,27 @@
 )
 
 
-# TODO: Add cluster parameters as kwargs here and add to task config
+def binarize_values(y_true, y_pred):
```
Comment (mlgill) [inline src/czbenchmarks/tasks/utils.py:28]: Could we clarify expected types here as np.arrays?

File: src/czbenchmarks/tasks/single_cell/perturbation_expression_prediction.py [hunk 10]
```diff
@@ -0,0 +1,300 @@
...
+        Returns:
+            PerturbationExpressionPredictionOutput: Output object containing dictionaries of predicted and true log fold changes
+            for each perturbation condition.
+        """
+        de_results = task_input.de_results
+        gene_to_ensembl = dict(
+            zip(de_results["condition"], de_results["condition_ensembl_id"])
+        )
+
+        samples = cell_representation.sample_id
```
Comment (mlgill) [inline src/czbenchmarks/tasks/single_cell/perturbation_expression_prediction.py:94]: What is `sample_id` if cell_representation is anndata?

File: examples/test_equivalency_perturbation_task.py (hunk 7 above)
Comment (polinabinder1) [reply examples/test_equivalency_perturbation_task.py:2]: That is accurate.

File: src/czbenchmarks/tasks/utils.py (hunk 9 above)
Comment (polinabinder1) [reply src/czbenchmarks/tasks/utils.py:28]: fixed.

File: src/czbenchmarks/metrics/implementations.py [hunk 11]
```diff
@@ -146,3 +166,43 @@
...
+    MetricType.ACCURACY,
+    func=accuracy_score,
+    required_args={"y_true", "y_pred"},
+    description="Accuracy between true and predicted values",
+    tags={"label_prediction"},
```
Comment (mlgill) [inline src/czbenchmarks/metrics/implementations.py:175]: It looks like we should update these tags where appropriate to add the perturb seq task.

File: src/czbenchmarks/tasks/single_cell/perturbation_expression_prediction.py [hunk 12]
```diff
@@ -0,0 +1,322 @@
...
+
+    def __init__(
+        self,
+        min_de_genes: int = 10,
+        control_gene: str = "non-targeting",
+        metric_column: str = "logfoldchange",
+        metric_type: str = "wilcoxon",
+        pval_threshold: float = 1e-4,
+        standardized_mean_diff: float = 0.5,
+        min_logfoldchange: float = 1.0,
```
Comment (mlgill) [inline src/czbenchmarks/tasks/single_cell/perturbation_expression_prediction.py:43]: With the exception of `metric_column`, there are (now) all saved within the single cell perturbation dataset. To ensure they are always set the same, I think they should probably move to `PerturbationExpressionPredictionTaskInput` to indicate they should be set by the dataset. Does that seem like it helps to clarify where they originate from?

File: examples/test_equivalency_perturbation_task.py [hunk 13]
```diff
@@ -0,0 +1,378 @@
+# This tests that the perturbation prediection task with the new data formats matches
```
Comment (mlgill) [inline examples/test_equivalency_perturbation_task.py:1]: Reviewer note: this is for internal testing and will be removed before merge

File: examples/test_equivalency_perturbation_dataset.py [hunk 14]
```diff
@@ -0,0 +1,325 @@
+import argparse
```
Comment (mlgill) [inline examples/test_equivalency_perturbation_dataset.py:1]: Reviewer note: this is for internal testing and will be removed before merge

File: src/czbenchmarks/tasks/single_cell/perturbation_expression_prediction.py [hunk 15]
```diff
@@ -0,0 +1,322 @@
...
+class PerturbationExpressionPredictionTaskInput(TaskInput):
+    """Pydantic model for PerturbationTask inputs."""
+
+    de_results: pd.DataFrame
+    masked_adata_obs: pd.DataFrame
+    var_index: pd.Index
+    target_genes_to_save: Dict[str, List[str]]
+
+
+class PerturbationExpressionPredictionOutput(TaskOutput):
```
Comment (atolopko-czi) [inline src/czbenchmarks/tasks/single_cell/perturbation_expression_prediction.py:25]: Just a nit and out-of-scope for this PR, but just realized all of the tasks' input types are named `*TaskInput` and all output types are named just `*Output`. Oh well. Worth a separate issue to fix this? 😄

File: .github/workflows/release-please.yaml [hunk 16]
```diff
@@ -21,8 +21,8 @@ jobs:
         id: generate_token
         uses: actions/create-github-app-token@v1
         with:
-          app-id: ${{ secrets.CZI_RELEASE_PLEASE_APP_ID }}
-          private-key: ${{ secrets.CZI_RELEASE_PLEASE_PK }}
+          app-id: ${{ secrets.GH_ACTIONS_HELPER_APP_ID }}
+          private-key: ${{ secrets.GH_ACTIONS_HELPER_PK }}
```
Comment (atolopko-czi) [inline .github/workflows/release-please.yaml:25]: did this break?

**Review by atolopko-czi — COMMENTED** [review]
Not quite a full review, so just a few comments I'll submit for now.

File: src/czbenchmarks/tasks/single_cell/perturbation_expression_prediction.py [hunk 17]
```diff
@@ -0,0 +1,300 @@
...
+                de_results["standardized_mean_diff"].abs()
+                >= self.standardized_mean_diff
+            ]
+
+        else:
+            raise ValueError(f"Metric type {self.metric_type} not supported")
+
+        for g_val, group in cell_representation.groupby("gene_condition", sort=False):
+            gene = gene_to_ensembl.get(g_val, g_val)
+            condition_de_df = de_results[de_results["condition_ensembl_id"] == gene]
```
[the remark this answers is not in the document]
Comment (polinabinder1) [reply src/czbenchmarks/tasks/single_cell/perturbation_expression_prediction.py]: updated

File: src/czbenchmarks/datasets/single_cell_perturbation.py [hunk 18]
```diff
@@ -365,6 +367,57 @@ def store_task_inputs(self) -> Path:
 
         return self.task_inputs_dir
 
+    def load_from_task_inputs(self, task_inputs_dir: Optional[Path] = None) -> None:
```
[the remark this answers is not in the document]
Comment (polinabinder1) [reply src/czbenchmarks/datasets/single_cell_perturbation.py:369]: The intended use case is reloading from the saved inputs. This would avoid needing to re-process the entire dataset.

File: src/czbenchmarks/metrics/implementations.py [hunk 19]
```diff
@@ -17,18 +17,38 @@
...
+from scipy.stats import pearsonr, spearmanr
 from .utils import compute_entropy_per_cell, mean_fold_metric, jaccard_score
 
 from .types import MetricRegistry, MetricType
 
 
+def spearman_correlation(a, b):
+    """Wrapper for spearmanr that returns only the correlation coefficient."""
+    result = spearmanr(a, b)
+    # Handle both old and new scipy versions
```
[the remark this answers is not in the document]
Comment (polinabinder1) [reply src/czbenchmarks/metrics/implementations.py:40]: That's a good idea! added.

File: .github/workflows/release-please.yaml (hunk 16 above)
Comment (polinabinder1) [reply .github/workflows/release-please.yaml:25]: I will revert these files - some things got swept up in the pre-commit run.

File: examples/using_czbenchmarks.ipynb [hunk 20]
```diff
@@ -47,7 +47,6 @@
    "metadata": {},
    "outputs": [],
    "source": [
-    "\n",
```
[the remark this answers is not in the document]
Comment (polinabinder1) [reply examples/using_czbenchmarks.ipynb:50]: these things were unintentionally added with pre-commit. I will revert them.

File: src/czbenchmarks/tasks/single_cell/k562_perturbation.py (hunk 5 above)
Comment (polinabinder1) [reply src/czbenchmarks/tasks/single_cell/k562_perturbation.py]: done

File: examples/test_equivalency_perturbation_task.py (hunk 13 above)
Comment (polinabinder1) [reply examples/test_equivalency_perturbation_task.py:1]: noted

File: src/czbenchmarks/datasets/single_cell_perturbation.py (hunk 18 above)
[the remark this answers is not in the document]
Comment (polinabinder1) [reply src/czbenchmarks/datasets/single_cell_perturbation.py:522]: Where would it be best to move this? I could add it as a standalone function for the dataset. I can also remove it for now and add it to a seperate PR later.

File: src/czbenchmarks/metrics/implementations.py (hunk 11 above)
Comment (polinabinder1) [reply src/czbenchmarks/metrics/implementations.py:175]: I have removed those

File: src/czbenchmarks/tasks/single_cell/perturbation_expression_prediction.py (hunk 15 above)
Comment (mlgill) [reply src/czbenchmarks/tasks/single_cell/perturbation_expression_prediction.py:29]: I made an issue here: https://github.com/chanzuckerberg/cz-benchmarks/issues/376

File: examples/perturbation_expression_prediction_task.py [hunk 21]
```diff
@@ -0,0 +1,92 @@
+import logging
```
[the remark this answers is not in the document]
Comment (polinabinder1) [reply examples/perturbation_expression_prediction_task.py:1]: will do!

**Review by atolopko-czi — APPROVED** [review]
Optimistically approving, since I realized the comments from my previous review are sufficiently minor and shouldn't require another review.

File: examples/example_perturbation_expression_prediction.py [hunk 22]
```diff
@@ -0,0 +1,142 @@
...
+    )
+    parser.add_argument(
+        "--save-inputs",
+        action="store_true",
+        help="Save dataset task inputs to disk and load them back (demonstrates save/load functionality)",
+    )
+    parser.add_argument(
+        "--metric",
+        type=str,
+        default="wilcoxon",  # wilcoxon or t-test, do not use t_test
```
[the remark this answers is not in the document]
Comment (mlgill) [reply examples/example_perturbation_expression_prediction.py:56]: We should remove the note about not using a t-test. :) This should probably say something like "Set this to correspond to the type of statistical test used to determine differentially expressed genes". The reason this script cares about which one was used is that it dicates which columns are used for a data filtering step in the dataset class.

The choice for which to use in the DEG analysis is more complex and has a dependency on the model output. If you're interested...Wilcoxon rank sum is preferred because it is parametric and, thus, doesn't require the data to be normally distributed. A t-test assumes a normal distribution, which has more statistical power than Wilcoxon rank sum. But this assumption can cause errors (primarily false positives) in the results, so Wilcoxon rank sum is usually preferred. However, there are a few models which require the use of a t-test due to normalizations they apply to the data (the case with one of our team's models). Thus we support both.

File: src/czbenchmarks/metrics/implementations.py [hunk 23]
```diff
@@ -146,3 +159,43 @@
     default_params={"metric": "auroc", "classifier": None},
     tags={"label_prediction"},
 )
+
+metrics_registry.register(
+    MetricType.ACCURACY,
```
Comment (steveherrin) [inline src/czbenchmarks/metrics/implementations.py:164]: We're going to get conflicts on these because I also had to add these: https://github.com/chanzuckerberg/cz-benchmarks/pull/380/files#diff-b4751e066d5363c5911ffa5ac2e71bef33cb4ee56bb48fdcd4e4a72478b44f72R116

But I implemented them by having them computed within the task, similar to how we're doing the mean-over-folds metrics, rather than requiring the task to return a full set of predictions

File: examples/example_perturbation_expression_prediction.py [hunk 24]
```diff
@@ -0,0 +1,142 @@
...
+                "min_smd": args.min_smd,
+            }
+        }
+    }
+
+    with tempfile.TemporaryDirectory() as d:
+        cfg_path = Path(d) / "config.yaml"
+        cfg_path.write_text(yaml.safe_dump(cfg))
+        dataset: SingleCellPerturbationDataset = load_dataset(
+            "replogle_k562_essential_perturbpredict", config_path=str(cfg_path)
```
[the remark this answers is not in the document]
Comment (polinabinder1) [reply examples/example_perturbation_expression_prediction.py:112]: added TODO
